File: FE/app/src/main/java/com/example/myapplication/model/Bike.java
package com.example.myapplication.model;

import com.google.gson.annotations.SerializedName;
import java.util.List;

public class Bike {
    @SerializedName("_id")
    private String id;
    
    private String name;
    private String brand;
    private String model;
    private double price;
    
    @SerializedName("originalPrice")
    private double originalPrice;
    
    private String description;
    private Specifications specifications;
    private List<Image> images;
    private List<Color> colors;
    private String category;
    private String status;
    private int stock;
    private List<String> features;
    private String warranty;
    private Rating rating;
    
    @SerializedName("isFeatured")
    private boolean featured;
    
    @SerializedName("isNew")
    private boolean isNew;
    
    private List<String> tags;
    
    @SerializedName("createdAt")
    private String createdAt;
    
    @SerializedName("updatedAt")
    private String updatedAt;
    
    @SerializedName("discountPercentage")
    private int discountPercentage;

    // Constructors
    public Bike() {}

    public Bike(String name, String brand, String model, double price, String description, String category) {
        this.name = name;
        this.brand = brand;
        this.model = model;
        this.price = price;
        this.description = description;
        this.category = category;
    }

    // Getters and Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getBrand() {
        return brand;
    }

    public void setBrand(String brand) {
        this.brand = brand;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public double getOriginalPrice() {
        return originalPrice;
    }

    public void setOriginalPrice(double originalPrice) {
        this.originalPrice = originalPrice;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Specifications getSpecifications() {
        return specifications;
    }

    public void setSpecifications(Specifications specifications) {
        this.specifications = specifications;
    }

    public List<Image> getImages() {
        return images;
    }

    public void setImages(List<Image> images) {
        this.images = images;
    }

    public List<Color> getColors() {
        return colors;
    }

    public void setColors(List<Color> colors) {
        this.colors = colors;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public int getStock() {
        return stock;
    }

    public void setStock(int stock) {
        this.stock = stock;
    }

    public List<String> getFeatures() {
        return features;
    }

    public void setFeatures(List<String> features) {
        this.features = features;
    }

    public String getWarranty() {
        return warranty;
    }

    public void setWarranty(String warranty) {
        this.warranty = warranty;
    }

    public Rating getRating() {
        return rating;
    }

    public void setRating(Rating rating) {
        this.rating = rating;
    }

    public boolean isFeatured() {
        return featured;
    }

    public void setFeatured(boolean featured) {
        this.featured = featured;
    }

    public boolean isNew() {
        return isNew;
    }

    public void setNew(boolean aNew) {
        isNew = aNew;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public String getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(String updatedAt) {
        this.updatedAt = updatedAt;
    }

    public int getDiscountPercentage() {
        return discountPercentage;
    }

    public void setDiscountPercentage(int discountPercentage) {
        this.discountPercentage = discountPercentage;
    }

    // Helper methods
    public String getFormattedPrice() {
        return String.format("%,.0f VND", price);
    }

    public String getFormattedOriginalPrice() {
        return String.format("%,.0f VND", originalPrice);
    }

    public String getCategoryDisplayName() {
        switch (category) {
            case "city": return "Xe đạp điện thành phố";
            case "mountain": return "Xe đạp điện leo núi";
            case "folding": return "Xe đạp điện gấp";
            case "cargo": return "Xe đạp điện chở hàng";
            case "sport": return "Xe đạp điện thể thao";
            default: return "Khác";
        }
    }

    public String getFirstImageUrl() {
        if (images != null && !images.isEmpty()) {
            return images.get(0).getUrl();
        }
        return null;
    }

    // Inner classes
    public static class Specifications {
        private String battery;
        private String motor;
        private String range;
        
        @SerializedName("maxSpeed")
        private String maxSpeed;
        
        private String weight;
        
        @SerializedName("chargingTime")
        private String chargingTime;

        // Getters and Setters
        public String getBattery() { return battery; }
        public void setBattery(String battery) { this.battery = battery; }
        public String getMotor() { return motor; }
        public void setMotor(String motor) { this.motor = motor; }
        public String getRange() { return range; }
        public void setRange(String range) { this.range = range; }
        public String getMaxSpeed() { return maxSpeed; }
        public void setMaxSpeed(String maxSpeed) { this.maxSpeed = maxSpeed; }
        public String getWeight() { return weight; }
        public void setWeight(String weight) { this.weight = weight; }
        public String getChargingTime() { return chargingTime; }
        public void setChargingTime(String chargingTime) { this.chargingTime = chargingTime; }
    }

    public static class Image {
        private String url;
        private String alt;

        public Image() {}
        public Image(String url, String alt) {
            this.url = url;
            this.alt = alt;
        }

        public String getUrl() { return url; }
        public void setUrl(String url) { this.url = url; }
        public String getAlt() { return alt; }
        public void setAlt(String alt) { this.alt = alt; }
    }

    public static class Color {
        private String name;
        private String hex;

        public Color() {}
        public Color(String name, String hex) {
            this.name = name;
            this.hex = hex;
        }

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        public String getHex() { return hex; }
        public void setHex(String hex) { this.hex = hex; }
    }

    public static class Rating {
        private double average;
        private int count;

        public Rating() {}
        public Rating(double average, int count) {
            this.average = average;
            this.count = count;
        }

        public double getAverage() { return average; }
        public void setAverage(double average) { this.average = average; }
        public int getCount() { return count; }
        public void setCount(int count) { this.count = count; }
    }
}

File: BE/models/Bike.js
const mongoose = require('mongoose');

const bikeSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Tên xe đạp điện là bắt buộc'],
    trim: true,
    maxlength: [100, 'Tên xe đạp điện không được vượt quá 100 ký tự']
  },
  brand: {
    type: String,
    required: [true, 'Thương hiệu là bắt buộc'],
    trim: true,
    maxlength: [50, 'Thương hiệu không được vượt quá 50 ký tự']
  },
  model: {
    type: String,
    required: [true, 'Model là bắt buộc'],
    trim: true,
    maxlength: [50, 'Model không được vượt quá 50 ký tự']
  },
  price: {
    type: Number,
    required: [true, 'Giá là bắt buộc'],
    min: [0, 'Giá không được âm']
  },
  originalPrice: {
    type: Number,
    min: [0, 'Giá gốc không được âm']
  },
  description: {
    type: String,
    required: [true, 'Mô tả là bắt buộc'],
    trim: true,
    maxlength: [1000, 'Mô tả không được vượt quá 1000 ký tự']
  },
  specifications: {
    battery: {
      type: String,
      trim: true
    },
    motor: {
      type: String,
      trim: true
    },
    range: {
      type: String,
      trim: true
    },
    maxSpeed: {
      type: String,
      trim: true
    },
    weight: {
      type: String,
      trim: true
    },
    chargingTime: {
      type: String,
      trim: true
    }
  },
  images: [{
    url: {
      type: String,
      required: true
    },
    alt: {
      type: String,
      default: ''
    }
  }],
  colors: [{
    name: {
      type: String,
      required: true
    },
    hex: {
      type: String,
      required: true
    }
  }],
  category: {
    type: String,
    required: [true, 'Danh mục là bắt buộc'],
    enum: ['city', 'mountain', 'folding', 'cargo', 'sport', 'other'],
    default: 'city'
  },
  status: {
    type: String,
    enum: ['available', 'out_of_stock', 'discontinued'],
    default: 'available'
  },
  stock: {
    type: Number,
    default: 0,
    min: [0, 'Số lượng tồn kho không được âm']
  },
  features: [{
    type: String,
    trim: true
  }],
  warranty: {
    type: String,
    trim: true,
    default: '12 tháng'
  },
  rating: {
    average: {
      type: Number,
      default: 0,
      min: 0,
      max: 5
    },
    count: {
      type: Number,
      default: 0
    }
  },
  isFeatured: {
    type: Boolean,
    default: false
  },
  isNew: {
    type: Boolean,
    default: true
  },
  tags: [{
    type: String,
    trim: true
  }]
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for discount percentage
bikeSchema.virtual('discountPercentage').get(function() {
  if (this.originalPrice && this.originalPrice > this.price) {
    return Math.round(((this.originalPrice - this.price) / this.originalPrice) * 100);
  }
  return 0;
});

// Index for better search performance
bikeSchema.index({ name: 'text', brand: 'text', model: 'text', description: 'text' });
bikeSchema.index({ category: 1, status: 1 });
bikeSchema.index({ price: 1 });
bikeSchema.index({ 'rating.average': -1 });

// Pre-save middleware
bikeSchema.pre('save', function(next) {
  // Auto-generate tags from name, brand, and category
  if (this.isModified('name') || this.isModified('brand') || this.isModified('category')) {
    this.tags = [
      ...this.name.toLowerCase().split(' '),
      this.brand.toLowerCase(),
      this.category.toLowerCase()
    ].filter(tag => tag.length > 2);
  }
  next();
});

module.exports = mongoose.model('Bike', bikeSchema);


